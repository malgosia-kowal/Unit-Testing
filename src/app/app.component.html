<div class="header">
  <h1>
    {{ title }}
  </h1>
</div>
<div class="container">
  <app-products></app-products>
  <app-quickview></app-quickview>
  <app-basket></app-basket>
</div>