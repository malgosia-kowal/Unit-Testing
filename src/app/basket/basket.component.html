<div class="basketContainer">
  <div class="basketContainerDetail">
    <div> Total: {{ basketService.total }} zl</div>
    <div> Products: {{ basketService.products.length }}</div>
  </div>
  <div>
    <app-button
      text="Clean"
      (action)="basketService.clear()"
    >
    </app-button>
  </div>
</div>
