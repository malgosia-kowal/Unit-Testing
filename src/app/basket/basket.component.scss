.basketContainer {
  z-index: 999;
  position: fixed;
  bottom: 0;
  background: white;
  width: 100vw;
  box-shadow: 0 0 10px rgb(0, 0, 0, 0.3);
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: 10px;

  &Detail {
    display: flex;
    flex-direction: column;
    justify-content: space-evenly;

    & > div {
      font-size: 20px;
    }
  }

  &Buttons {
    display: flex;
    flex-direction: row;
    margin-right: 40px;
  }
}
