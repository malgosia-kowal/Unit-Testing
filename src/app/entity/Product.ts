export class Product {
  public name: string;
  public price: number;
  public quantity: number;
  public size: string;
  public image: string;

  constructor(name: string, price: number, quantity: number, size: string, image: string) {
    this.name = name;
    this.price = price;
    this.quantity = quantity;
    this.size = size;
    this.image = image; 
  }
}
