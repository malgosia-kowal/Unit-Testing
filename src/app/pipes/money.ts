import { Pipe, PipeTransform } from "@angular/core";

@Pipe({ name: "money" })
export class MoneyPipe implements PipeTransform {
  transform(value: number): string {
    return (value / 100).toFixed(2).replace(".", ",");
  }
}
