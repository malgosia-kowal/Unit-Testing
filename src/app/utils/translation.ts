import { TranslateService } from "@ngx-translate/core";
import { Locale } from "../app.component";

export const getPreviousLang = (
  translate: TranslateService,
  lang: string
): Locale => {
  return (translate.getLangs().filter(l => l !== lang)[0] ||
    translate.getDefaultLang()) as Locale;
};
