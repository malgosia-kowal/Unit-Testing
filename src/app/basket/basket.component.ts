import { Component, OnInit } from '@angular/core';
import { BasketService } from '../service/basket.service';
import { Toggable, ToggleService } from '../service/toggle.service';

@Component({
  selector: 'app-basket',
  templateUrl: './basket.component.html',
  styleUrls: ['./basket.component.scss']
})
export class BasketComponent implements OnInit {

  constructor(public basketService: BasketService, public toggleService: ToggleService) { }

  ngOnInit() {
  }

  toggleQuickview() {
    this.toggleService.toggle(Toggable.Quickview);
  }

}
