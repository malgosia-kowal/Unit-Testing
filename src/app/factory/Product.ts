import { Product } from '../entity/Product'; 
import * as faker from 'faker';

export const createProduct = (data: Partial<Product> = {}) => new Product(
    data.hasOwnProperty('name') ? data.name : faker.commerce.productName(), 
    data.hasOwnProperty('price') ? data.price : parseInt(faker.commerce.price()), 
    1, 
    data.hasOwnProperty('size') ? data.size : faker.commerce.color(),
    data.hasOwnProperty('image') ? data.image : faker.image.imageUrl(),
);
