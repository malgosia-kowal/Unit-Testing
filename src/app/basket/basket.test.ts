import { async, ComponentFixture, TestBed } from "@angular/core/testing";
import { By } from "@angular/platform-browser";
import { BasketComponent } from "./basket.component";
import { ButtonComponent } from "../button/button.component";
import { BasketService } from "../service/basket.service";
import { ToggleService, Toggable } from "../service/toggle.service";
import { createProduct } from "../factory/Product";
import { BehaviorSubject } from "rxjs";
import {
  TranslateModule,
  TranslateLoader,
  TranslateService
} from "@ngx-translate/core";
import { Locale, AppComponent } from "../app.component";
import { MoneyPipe } from "../pipes/money";
import currencyService from "../service/currency.service";
import { MockCustomLoader } from "../tests/mocks/translation";


jest.mock("../service/currency.service");

describe("BasketComponent", () => {
  let component: BasketComponent;
  let fixture: ComponentFixture<BasketComponent>;
  let toggleService: ToggleService;
  let basketService: BasketService;
  let translate: TranslateService;

  beforeEach(async(async () => {
    TestBed.configureTestingModule({
      declarations: [BasketComponent, ButtonComponent, MoneyPipe],
      imports: [
        TranslateModule.forRoot({
          loader: { provide: TranslateLoader, useClass: MockCustomLoader }
        })
      ],
      providers: [
        { provide: ToggleService, useClass: ToggleService },
        { provide: BasketService, useClass: BasketService }
      ]
    })
      .compileComponents()
      .then(() => {
        fixture = TestBed.createComponent(BasketComponent);
        component = fixture.componentInstance;

        toggleService = TestBed.get(ToggleService);
        basketService = TestBed.get(BasketService);
        translate = TestBed.get(TranslateService);
        translate.use(Locale.Gb);

        fixture.detectChanges();
      });
    (currencyService.convert as jest.Mock).mockImplementation(
      () => new Promise(resolve => resolve(0))
    );
  }));

  it("should exist", () => {
    expect(component).toBeDefined();
  });

  it("should render correctly", () => {
    expect(fixture).toMatchSnapshot();
  });

  it("should render correctly with cat in the basket", async () => {
    const product = createProduct({ price: 320 });
    component.basketService.addProduct(product);

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it("should set the dependecies", () => {
    component.ngOnInit();

    expect(component.toggleService).toBe(toggleService);
    expect(component.basketService).toBe(basketService);
  });

  it("can toggle quickview", async () => {
    const toggleButton = fixture.debugElement.query(By.css("#quickViewButton"));

    toggleButton.triggerEventHandler("action", "test");

    const visibleSubject = toggleService.visible(Toggable.Quickview)
      .source as BehaviorSubject<boolean>;

    expect(visibleSubject.getValue()).toBeTruthy();

    toggleButton.triggerEventHandler("action", "test");
    expect(visibleSubject.getValue()).toBeFalsy();
  });

  it("can clean the basket", () => {
    const product = createProduct();
    component.basketService.addProduct(product);
    fixture.detectChanges();
    const cleanBasketButton = fixture.debugElement.query(
      By.css("#clearTheBasketButton")
    );

    cleanBasketButton.triggerEventHandler("action", "test");

    expect(component.basketService.products.getValue().length).toEqual(0);
  });
});
