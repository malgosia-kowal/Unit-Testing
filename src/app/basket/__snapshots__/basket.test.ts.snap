// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasketComponent should render correctly 1`] = `
<app-basket
  basketService={[Function BasketService]}
  toggleService={[Function ToggleService]}
>
  <div
    class="basketContainer"
  >
    <div
      class="basketContainerDetail"
    >
      <div>
        <span
          id="basketTotal"
          ng-reflect-translate="Basket.Total"
          ng-reflect-translate-params="[object Object]"
        >
          Total: 0,00 gbp
        </span>
      </div>
      <div>
        <span
          id="basketLength"
          ng-reflect-translate="Basket.Items"
          ng-reflect-translate-params="[object Object]"
        >
          Cats: 0
        </span>
      </div>
    </div>
    <div
      class="basketContainerButtons"
    >
      <app-button
        id="clearTheBasketButton"
        ng-reflect-text="Clean"
      >
        <button>
           Clean
  
        </button>
      </app-button>
      <app-button
        id="quickViewButton"
        ng-reflect-text="Quick view"
      >
        <button>
           Quick view
  
        </button>
      </app-button>
    </div>
  </div>
</app-basket>
`;

exports[`BasketComponent should render correctly with cat in the basket 1`] = `
<app-basket
  basketService={[Function BasketService]}
  toggleService={[Function ToggleService]}
>
  <div
    class="basketContainer"
  >
    <div
      class="basketContainerDetail"
    >
      <div>
        <span
          id="basketTotal"
          ng-reflect-translate="Basket.Total"
          ng-reflect-translate-params="[object Object]"
        >
          Total: 3,20 gbp
        </span>
      </div>
      <div>
        <span
          id="basketLength"
          ng-reflect-translate="Basket.Items"
          ng-reflect-translate-params="[object Object]"
        >
          Cats: 1
        </span>
      </div>
    </div>
    <div
      class="basketContainerButtons"
    >
      <app-button
        id="clearTheBasketButton"
        ng-reflect-text="Clean"
      >
        <button>
           Clean
  
        </button>
      </app-button>
      <app-button
        id="quickViewButton"
        ng-reflect-text="Quick view"
      >
        <button>
           Quick view
  
        </button>
      </app-button>
    </div>
  </div>
</app-basket>
`;
